artichoke-backend/src/extn/core/kernel/kernel.rb

```diff
@@ -40,6 +40,221 @@ def catch(tag = nil)
     e.value
   end
 
+  def Array(arg) # rubocop:disable Naming/MethodName
+    return arg if arg.is_a?(Array)
+
+    ret = nil
+    ret = arg.to_ary if arg.respond_to?(:to_ary)
+    ret = arg.to_a if ret.nil? && arg.respond_to?(:to_a)
+
+    return [arg] if ret.nil?
+    return ret if ret.is_a?(Array)
+
+    raise TypeError
+  end
+
+  def Hash(arg) # rubocop:disable Naming/MethodName
+    return arg if arg.is_a?(Hash)
+    return {} if arg.nil? || arg == []
+
+    if arg.respond_to?(:to_hash)
+      ret = arg.to_hash
+      return {} if ret.nil? || ret.empty?
+      return ret if ret.is_a?(Hash)
+    end
+
+    raise TypeError
+  end
+
+  def Integer(arg, base = 0, exception: true) # rubocop:disable Naming/MethodName
+    raise ArgumentError, 'base specified for non string value' if base.positive? && arg.is_a?(Numeric)
+
+    classname = arg.class
+    classname = arg.inspect if arg.nil? || arg.equal?(false) || arg.equal?(true)
+
+    if arg&.respond_to?(:to_int)
+      ret = arg.to_int
+
+      # uncritically return the value of to_int even if it is not an Integer
+      return ret if ret.is_a?(Numeric)
+      return ret if ret&.scan(/\D/)&.empty?
+
+      ret = arg.to_i
+
+      return ret if ret.is_a?(Numeric)
+
+      raise TypeError, "can't convert #{classname} to Integer (#{arg.class}#to_i gives #{ret.class})"
+    elsif !arg.is_a?(String) && arg && arg.respond_to?(:to_i)
+      ret = arg.to_i
+
+      return ret if ret.is_a?(Numeric)
+
+      raise TypeError, "can't convert #{classname} to Integer (#{arg.class}#to_i gives #{ret.class})"
+    elsif arg.is_a?(String)
+      i = 0
+
+      # squeeze preceding spaces
+      i += 1 while arg[i] == ' '
+
+      # handle sign
+      sign = 1
+      if arg[i] == '+'
+        i += 1
+      elsif arg[i] == '-'
+        i += 1
+        sign = -1
+      end
+
+      # no multiple leading +/-, i.e. "++1", "---1"
+      # no space between -/+ and next digit, i.e. "+ 1"
+      # no leading _
+      # TODO: no leading null byte \0
+      raise ArgumentError, "invalid value for Integer(): \"#{arg}\"" if arg[i] == '+' || arg[i] == '-' || arg[i] == ' ' || arg[i] == '_'
+
+      # no multiple embedded _
+      # TODO: no null bytel \0
+      j = i
+      while j < arg.length
+        raise ArgumentError, "invalid value for Integer(): \"#{arg}\"" if arg[j] == '_' && arg[j + 1] == '_'
+
+        j += 1
+      end
+
+      # no trailing -, +, i.e. "1-", "1++"
+      # no trailing _
+      # TODO: no trailing null byte \0
+      k = arg.length - 1
+      # squeeze trailing spaces
+      k += 1 while arg[k] == ' '
+      raise ArgumentError, "invalid value for Integer(): \"#{arg}\"" if arg[k] == '-' || arg[k] == '+' || arg[k] == '_'
+
+      # when base is not specified or zero, check the base letter
+      if base <= 0
```

Comment:
Oh no 😱 This looks very complicated and hard to get right.

Now I understand why you were asking about the split between Rust and Ruby. In this case, Rust has a library function that will do all of this hard work for us in a much better tested way: [`i64::from_str_radix`](https://doc.rust-lang.org/std/primitive.i64.html#method.from_str_radix).

I've mocked up an example parsing function here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=65200add357dd7d7d87b15e44db8dd62